### ISSUE TITLE
Opentsdb: Support for Opentsdb 2.2 filters

### ISSUE THREAD

**Issue by utkarshcmu** [grafana/grafana#3077]
Opentsdb 2.2 now supports filters which is a significant improvement in filtering the queries. It did not have wildcard support etc. Getting filters in Grafana will make Grafana so powerful. This will be like having functions for Graphite. 

http://opentsdb.net/docs/build/html/user_guide/query/filters.html

**Comment by torkelo** [comment]
Not sure it equates to graphite functions at all, these filters  offer no metrics transformations at all just new ways to select metrics by tag values, so more akin to Graphite glob expressions and wildcard patterns. 

Anyway agree, it looks useful. Not sure what the best way to add support for it is, oris anything needed at all?

**Comment by utkarshcmu** [comment]
Yes, its not exactly like Graphite functions but this will be very useful in filtering queries. For example,

Opentsdb Tag Key - server
Opentsdb Tag Values - a1,a2,a3,a4,a5,b1,b2,b3,b4,b5

Now, we can select metrics without filters in 3 ways:
1. key= server, value= *
2. key= server, value= a1|b3
3. key= server, value= b2

There is no way I can write the following combination in Opentsdb 2.1:
1. key= server, value=a*  --> This won't work in Opentsdb in 2.1

Now, after reading the filters doc, in Opentsdb 2.2 we can run the above query and it should be fine. 

Will investigate and update this ticket if anything has to be done for supporting filters from Grafana's end. If nothing has to be done then we can close the ticket. Will update very soon.

**Comment by utkarshcmu** [comment]
We do not need to do anything for supporting filters on Grafana's side, it is supported via Opentsdb in v2.2.0. You may close this Issue. :)

**Comment by dentcat** [comment]
OpenTSDB 2.2 has query filter support but does Grafana query editor for OpenTSDB support it?

I can't find a way to specify filter type or groupBy attribute in the Grafana UI.

**Comment by utkarshcmu** [comment]
Will look into it and get back asap. Thanks for reporting this.

**Comment by utkarshcmu** [comment]
Need to edit the query control for Opentsdb for supporting filters, here is the reference:
http://opentsdb.net/docs/build/html/api_http/query/index.html

The above doc was not updated when I checked so that was the reason for closing this ticket. Reopening it now.

@torkelo This issue should be high priority for Opentsdb datasource users as people have been waiting for filters since a very long time. Will work on the PR and submit it asap. :)

**Comment by utkarshcmu** [comment]
Also, tags are deprecated in 2.2, so will be making the changes in PR #3234.

**Comment by digrich** [comment]
+1

**Comment by johann8384** [comment]
The company I worked for submitted the expression code and has an internal Datasource that works with the expressions, although, it's simplistic at best. It is simply a text box where you write your expression to be returned.

I'm planning to submit a PR to add that datasource, but if @utkarshcmu has something better in the works, like improvements to the OpenTSDB datasource to support them the way the Graphite or Bosun datasources do, that would be great.

An example of the new expression usage:

```bash
curl 'http://localhost:4244/api/query/gexp/?start=10m-ago&exp=alias(scale(sum:rate:1m-avg:tsd.rpc.received\{domain=atl1\},0.10),AdsReturned)' -H 'Accept: application/json, text/plain, */*' -H 'Referer: http://localhost/grafana/' -H 'Origin: http://localhost' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.86 Safari/537.36' --compressed
```

A picture of our Datasource in Use:

<img width="400" alt="grafana-opentsdb-expression" src="https://cloud.githubusercontent.com/assets/154673/12455365/017cc104-bf61-11e5-9fdc-266c1da88674.png">

**Comment by amalshah** [comment]
+1 for supporting the groupBy clause in OpenTSDB queries.
I'm using Grafana 2.6 and OpenTSDB 2.2 RC3. 

I'm trying to aggregate certain set of tag values and show summarized graphs on the aggregations.
So, if "my.metric" has a tag "my.tag" with values "a.1", "a.2", "b.1", "b.2", I'd like to show aggregate graphs for "a.*" and "b.*".
Retrieving the data sets using wildcards works ok. So, I have  
Tag: my.tag = wildcard(a.*) 
However, this retrieves a.1, a.2 as separate timeseries, as by default it seems that the groupBy clause sent to OpenTSDB is true.
Having a checkbox in the OpenTSDB query panel, groupBy=true|false, would allow one to either see separate timeseries or have them aggregated.

**Comment by pilgrimkst** [comment]
+1 for group by

**Comment by utkarshcmu** [comment]
I am working on supporting features , will be submitting a PR very soon:
Here is the progress of the work: https://github.com/utkarshcmu/grafana/tree/tsdb-refac

**Comment by amalshah** [comment]
I did make some local changes to get this working, as attached.
Not fully tested, or complete, and based on my very limited understanding of the source, but may be useful to you. The changes are marked by: *FILTERSUPPORT* in opentsdb/datasource.js and the query.editor.html file.

**Comment by bergquist** [comment]
solved by #4138
